### PR TITLE
feat: Combine gas estimate UI with router changes

### PR DESCRIPTION
**Pull request by ianlapham**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -173,5 +173,7 @@
     ]
   },
   "license": "GPL-3.0-or-later",
-  "dependencies": {}
+  "dependencies": {
+    "use-resize-observer": "^8.0.0"
```
Comment (justindomingue) [inline package.json:177]: for consistency might want to move into this to dev dependencies

File: src/components/CurrencyInputPanel/FiatValue.tsx [hunk 2]
```diff
@@ -25,18 +26,24 @@ export function FiatValue({
...
+          />
         </Trans>
       ) : (
         ''
       )}
       {priceImpact ? (
         <span style={{ color: priceImpactColor }}>
           {' '}
-          (<Trans>{priceImpact.multiply(-1).toSignificant(3)}%</Trans>)
+          <MouseoverTooltip text={'The estimated difference between the USD values of input and output amounts.'}>
```
Comment (justindomingue) [inline src/components/CurrencyInputPanel/FiatValue.tsx:44]: don't forget to translate, can use `t` from `const { t } = useTranslation()` I think

File: src/components/CurrencyInputPanel/index.tsx [hunk 3]
```diff
@@ -29,15 +29,17 @@ const InputPanel = styled.div<{ hideInput?: boolean }>`
...
 
 const FixedContainer = styled.div`
   width: 100%;
   height: 100%;
   position: absolute;
   border-radius: 20px;
-  background-color: ${({ theme }) => theme.bg1};
-  opacity: 0.95;
+  background-color: ${({ theme }) => theme.bg2};
+  // opacity: 0.95;
```
Comment (justindomingue) [inline src/components/CurrencyInputPanel/index.tsx:42]: remove comment?

File: src/components/Header/Polling.tsx [hunk 4]
```diff
@@ -43,6 +57,17 @@ const StyledPollingDot = styled.div<{ warning: boolean }>`
   transition: 250ms ease background-color;
 `
 
+const StyledGasDot = styled.div`
+  width: 4px;
```
Comment (justindomingue) [inline src/components/Header/Polling.tsx:61]: we're not super good about this but we usually try to keep the properties sorted

File: src/components/Header/Polling.tsx [hunk 5]
```diff
@@ -105,19 +134,52 @@ export default function Polling() {
...
+                <TYPE.main fontSize="11px" mr="8px" color={theme.text3}>
+                  <MouseoverTooltip
+                    text={
+                      <Trans>
+                        {`The current fast gas amount for sending a transaction on L1.
+                    Gas fees are paid in Ethereum's native currency Ether (ETH) and denominated in gwei. `}
+                      </Trans>
+                    }
+                  >
+                    {priceGwei.toString()} gwei
```
Comment (justindomingue) [inline src/components/Header/Polling.tsx:159]: hm probably need this in a `<Trans>` for languages that change the order

File: src/components/Identicon/index.tsx [hunk 6]
```diff
@@ -31,7 +31,11 @@ export default function Identicon() {
       const current = ref.current
       current?.appendChild(icon)
       return () => {
-        current?.removeChild(icon)
+        try {
+          current?.removeChild(icon)
```
Comment (justindomingue) [inline src/components/Identicon/index.tsx:35]: this could probably be pushed now as a separate pr?

File: src/state/routing/types.ts [hunk 7]
```diff
@@ -62,3 +65,32 @@ export interface GetQuoteResult {
   route: Array<V3PoolInRoute[] | V2PoolInRoute[]>
   routeString: string
 }
+
+export class InterfaceTrade<
```
Comment (justindomingue) [inline src/state/routing/types.ts:69]: after talking with pote, I think we prefer to use composition over inheritance here, meaning

```
export type TradeWithGasEstimates = {
  trade: Trade<>,
  gasEstimates: {
    gasUseEstimateUSD,
  }
}
```

or something like that. WDYT?

File: src/components/CurrencyInputPanel/FiatValue.tsx (hunk 2 above)
Comment (ianlapham) [reply src/components/CurrencyInputPanel/FiatValue.tsx:44]: Looks like we dont have this hook anymore?

File: src/components/Header/Polling.tsx (hunk 4 above)
Comment (ianlapham) [reply src/components/Header/Polling.tsx:61]: Do we have a guideline for the ordering?

File: src/state/routing/types.ts (hunk 7 above)
Comment (ianlapham) [reply src/state/routing/types.ts:69]: First off, like this name better. 

Yeah im okay using composition - probably makes it easier if `Trade` is updated in future

File: src/components/AnimatedDropdown/index.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { animated, useSpring } from 'react-spring'
+import useResizeObserver from 'use-resize-observer'
+
+/**
+ * @param open conditional to show content or hide
+ * @returns Wrapper to smoothly hide and expand content
+ */
+export default function AnimatedDropdown({ open, children }: { open: boolean; children: JSX.Element }) {
```
Comment (justindomingue) [inline src/components/AnimatedDropdown/index.tsx:8]:
```suggestion
export default function AnimatedDropdown({ open, children }: React.PropsWithChildren<{ open: boolean }) {
```

can simplify type

File: src/components/AnimatedDropdown/index.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  const { ref, height } = useResizeObserver()
+
+  const props = useSpring({
+    maxHeight: open ? height ?? 0 : 0,
```
Comment (justindomingue) [inline src/components/AnimatedDropdown/index.tsx:12]: did you mean to keep this as `height`?

File: src/components/CurrencyInputPanel/FiatValue.tsx (hunk 2 above)
Comment (justindomingue) [reply src/components/CurrencyInputPanel/FiatValue.tsx:44]: hm so ideally you could use `<Trans>` here. However mouseovertooltip requires a `string` (i'm assuming, didn't actually check) so you have to use `t`

File: src/components/CurrencyInputPanel/index.tsx [hunk 10]
```diff
@@ -130,26 +131,24 @@ const StyledBalanceMax = styled.button<{ disabled?: boolean }>`
   background-color: transparent;
   border: none;
   border-radius: 12px;
-  font-size: 14px;
+  font-size: 11px;
   font-weight: 500;
   cursor: pointer;
-  padding: 0;
-  color: ${({ theme }) => theme.primaryText1};
+  padding: 4px 6px;
```
Comment (justindomingue) [inline src/components/CurrencyInputPanel/index.tsx:137]: nit. sort

File: src/components/Header/Polling.tsx (hunk 4 above)
Comment (justindomingue) [reply src/components/Header/Polling.tsx:61]: just alphabetical

File: src/components/Header/Polling.tsx [hunk 11]
```diff
@@ -81,6 +106,10 @@ export default function Polling() {
   const [isHover, setIsHover] = useState(false)
   const machineTime = useMachineTimeMs(NETWORK_HEALTH_CHECK_MS)
   const blockTime = useCurrentBlockTimestamp()
+  const theme = useTheme()
+
+  const ethGasPrice = useGasPrice()
+  const priceGwei = ethGasPrice ? JSBI.divide(ethGasPrice, JSBI.BigInt(1000000000)) : undefined
```
Comment (justindomingue) [inline src/components/Header/Polling.tsx:112]:
```suggestion
  const priceGwei = ethGasPrice ? JSBI.divide(ethGasPrice, JSBI.BigInt(1_000_000_000)) : undefined
```

Comment (justindomingue) [reply src/components/Header/Polling.tsx:112]: should confirm all browsers we target support this or that it gets removed at compilation

Comment (justindomingue) [reply src/components/Header/Polling.tsx:112]: could also use `1e9` for readability?

File: src/components/Header/Polling.tsx [hunk 12]
```diff
@@ -105,19 +134,48 @@ export default function Polling() {
...
+          </ExternalLink>
           <StyledPollingNumber breathe={isMounting} hovering={isHover}>
-            {blockNumber}&ensp;
+            <ExternalLink
+              href={
+                chainId && blockNumber ? getExplorerLink(chainId, blockNumber.toString(), ExplorerDataType.BLOCK) : ''
+              }
+            >
+              <MouseoverTooltip
+                text={<Trans>{`The most recent block number on this network. Prices update on every block.`}</Trans>}
```
Comment (justindomingue) [inline src/components/Header/Polling.tsx:169]: oh so mouseovertooltip.text does support a react node. Use `<Trans>` above too 👍

File: src/components/RoutingDiagram/RoutingDiagram.tsx [hunk 13]
```diff
@@ -99,9 +99,9 @@ export default function RoutingDiagram({
     <Wrapper>
       {routes.map((entry, index) => (
         <RouteContainerRow key={index}>
-          <CurrencyLogo currency={tokenIn} />
+          <CurrencyLogo currency={tokenIn} size={'20px'} />
```
Comment (justindomingue) [inline src/components/RoutingDiagram/RoutingDiagram.tsx:102]: wdyt of passing a number here instead `size={20}`?

File: src/components/RoutingDiagram/RoutingDiagram.tsx [hunk 14]
```diff
@@ -136,12 +136,18 @@ function Pool({ currency0, currency1, feeAmount }: { currency0: Currency; curren
...
+
   return (
-    <PoolBadge>
-      <Box margin="0 5px 0 10px">
-        <DoubleCurrencyLogo currency0={tokenInfo1} currency1={tokenInfo0} size={20} />
-      </Box>
-      <ThemedText.Small fontSize={12}>{feeAmount / 10000}%</ThemedText.Small>
-    </PoolBadge>
+    <MouseoverTooltip
+      text={<Trans>{tokenInfo0?.symbol + '/' + tokenInfo1?.symbol + ' ' + feeAmount / 10000}% pool</Trans>}
```
[the remark this answers is not in the document]
Comment (justindomingue) [reply src/components/RoutingDiagram/RoutingDiagram.tsx:143]: i think it's important to split up the variables here so the keys get properly extracted for translation

[the remark this answers is not in the document]
Comment (justindomingue) [reply src/components/RoutingDiagram/RoutingDiagram.tsx:143]: rn i think your key will be `{0}% pool` and that's it. Some languages may need fee amount to be before/after, etc. so my suggestion would extract `{0}/{1} {2}% pool`

File: src/components/Tooltip/index.tsx [hunk 15]
```diff
@@ -25,6 +25,7 @@ interface TooltipContentProps extends Omit<PopoverProps, 'content'> {
   onOpen?: () => void
   // whether to wrap the content in a `TooltipContainer`
   wrap?: boolean
+  disableHover?: boolean // disable the hover and content display
```
Comment (justindomingue) [inline src/components/Tooltip/index.tsx:28]: have you considered doing this via `content` instead of a new prop? i.e. caller would handle the disabled state and send `content=null` when it's disabled. 

not saying it's right approach, just wondering

File: src/theme/components.tsx [hunk 16]
```diff
@@ -261,3 +261,9 @@ export const SmallOnly = styled.span`
     display: block;
   `};
 `
+
+export const Separator = styled.div`
+  width: 100%;
```
Comment (justindomingue) [inline src/theme/components.tsx:266]: nit. sort

File: src/hooks/useUSDCPrice.ts [hunk 17]
```diff
@@ -66,3 +71,54 @@ export function useUSDCValue(currencyAmount: CurrencyAmount<Currency> | undefine
...
+export function useDefaultGasCostEstimate(): {
+  cost: CurrencyAmount<Token> | null | undefined
+  loading: boolean
+  syncing: boolean
+} {
+  const { chainId } = useActiveWeb3React()
+  const stablecoin = chainId ? STABLECOIN_AMOUNT_OUT[chainId]?.currency : undefined
+  const ether = useCurrency('ETH')
+  const dummyEthAmount = ether ? CurrencyAmount.fromRawAmount(ether, JSBI.BigInt(1)) : undefined
+  const trade = useBestTrade(TradeType.EXACT_INPUT, dummyEthAmount ?? undefined, stablecoin)
```
[the remark this answers is not in the document]
Comment (justindomingue) [reply src/hooks/useUSDCPrice.ts:112]: once concern is this will call the routing api every 10s which pote won't love 😄 

@willpote have thoughts here?

[the remark this answers is not in the document]
Comment (justindomingue) [reply src/hooks/useUSDCPrice.ts:112]: discussed offline, removing

File: src/hooks/useSwapSlippageTolerance.ts [hunk 18]
```diff
@@ -18,7 +19,7 @@ const ONE_TENTHS_PERCENT = new Percent(10, 10_000) // .10%
  * Return a guess of the gas cost used in computing slippage tolerance for a given trade
  * @param trade the trade for which to _guess_ the amount of gas it would cost to execute
  */
-function guesstimateGas(trade: Trade<Currency, Currency, TradeType> | undefined): number | undefined {
+export function guesstimateGas(trade: Trade<Currency, Currency, TradeType> | undefined): number | undefined {
```
Comment (justindomingue) [inline src/hooks/useSwapSlippageTolerance.ts:22]: does it need to be exported?

File: src/components/swap/SwapWarningDropdown.tsx [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+  fiatValueInput,
+  trade,
+}: {
+  fiatValueInput: CurrencyAmount<Token> | null
+  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined
+}) {
+  // gas cost estimate is more than half of input value
+  const showNetworkFeeWarning = Boolean(
+    fiatValueInput &&
+      trade &&
```
Comment (justindomingue) [inline src/components/swap/SwapWarningDropdown.tsx:38]: can simplify

```suggestion
      && trade?.gasUseEstimateUSD
```

File: src/components/swap/SwapRoute.tsx [hunk 20]
```diff
@@ -1,85 +1,97 @@
...
             ) : (
-              <Trans>
-                Best trade via {routes.length} routes on Uniswap {hasV2Routes ? 'V2' : 'V3'}
-              </Trans>
-            )}
-          </ThemedText.Main>
-        ))}
-    </AutoColumn>
+              <ThemedText.Main fontSize={12} width={400} margin={0}>
+                {/* could not get <Plural> to render `one` correctly. */}
```
Comment (justindomingue) [inline src/components/swap/SwapRoute.tsx:85]: yay no more plural, can remove this comment

File: src/components/swap/SwapRoute.tsx [hunk 21]
```diff
@@ -1,85 +1,97 @@
...
-  syncing,
-}: {
-  trade: Trade<Currency, Currency, TradeType>
+interface SwapRouteProps extends React.HTMLAttributes<HTMLDivElement> {
+  trade: InterfaceTrade<Currency, Currency, TradeType>
   syncing: boolean
-}) {
-  const autoRouterSupported = useAutoRouterSupported()
+  fixedOpen?: boolean // fixed in open state, hide open/close icon
+}
```
Comment (justindomingue) [inline src/components/swap/SwapRoute.tsx:46]: i think i could use composition over inheritance here 🤔 

```
type SwapRouteProps = {
trade :...
syncing:...
fixeOpen?: ...
} & React.HTMLAttributes<HTMLDivElement>
```

File: src/components/swap/GasEstimateBadge.tsx [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+  }
+`
+
+export default function GasEstimateBadge({
+  trade,
+  loading,
+  showRoute,
+  disableHover,
+}: {
+  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined | null // dollar amount in active chain's stabelcoin
```
Comment (justindomingue) [inline src/components/swap/GasEstimateBadge.tsx:41]:
```suggestion
  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined | null // dollar amount in active chain's stablecoin
```

File: src/components/swap/GasEstimateBadge.tsx [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+import { LoadingOpacityContainer } from 'components/Loader/styled'
+import { RowFixed } from 'components/Row'
+import { MouseoverTooltipContent } from 'components/Tooltip'
+import ReactGA from 'react-ga'
+import { InterfaceTrade } from 'state/routing/types'
+import styled from 'styled-components/macro'
+import { ThemedText } from 'theme'
+
+import { ReactComponent as GasIcon } from '../../assets/images/gas-icon.svg'
+// import { ReactComponent as GasIcon } from '../../assets/images/router-icon-grey.svg'
```
Comment (justindomingue) [inline src/components/swap/GasEstimateBadge.tsx:13]: remove?

File: src/components/AnimatedDropdown/index.tsx (hunk 9 above)
Comment (ianlapham) [reply src/components/AnimatedDropdown/index.tsx:12]: Yep good call

File: src/state/routing/slice.ts [hunk 24]
```diff
@@ -99,6 +108,7 @@ export const routingApi = createApi({
               tokenOutAddress,
               tokenOutChainId,
               amount,
+              gasPriceWei,
```
Comment (justindomingue) [inline src/state/routing/slice.ts:111]: do we actually use `gasPriceWei`?

Comment (ianlapham) [reply src/state/routing/slice.ts:111]: ah no this is leftover rn ill remove
